### ISSUE TITLE
Wrong filename hash when upcase table is compressed

### ISSUE THREAD

**Issue by eiselekd** [relan/exfat#30]
The upcase table can be compressed. When this is the case
the upcase conversion will not work and file hashes will not be compatible
with a driver that handles the compressed upcase table correctly.
Something like the below patch will fix it:

```

--- fuse-exfat-1.1.0/libexfat/node.c	2014-07-08 15:40:12.000000000 +0200
+++ fuse-exfat-1.1.0.ch/libexfat/node.c	2016-02-08 19:34:37.265341043 +0100
@@ -247,6 +247,7 @@
 	uint16_t reference_checksum = 0;
 	uint16_t actual_checksum = 0;
 	uint64_t valid_size = 0;
+	int i, j, k, ucsize;
 
 	*node = NULL;
 
@@ -398,6 +399,35 @@
 						le32_to_cpu(upcase->start_cluster));
 				goto error;
 			}
+			/* handle compressed upcase table */
+			for (ucsize = 0, i = 0; i < ef->upcase_chars; i++) {
+				if (le16_to_cpu(ef->upcase[i]) != 0xffff ||
+				    (i+1 >= ef->upcase_chars))
+					ucsize++;
+				else {
+					ucsize += le16_to_cpu(ef->upcase[i+1]);
+					i++;
+				}
+			}
+			if (ucsize <= 0x10000) {
+				le16_t *u =  (le16_t *)malloc(ucsize * sizeof(le16_t));
+				for (i = 0, j = 0; i < ef->upcase_chars; i++) {
+					if (le16_to_cpu(ef->upcase[i]) != 0xffff ||
+					    (i+1 >= ef->upcase_chars)) {
+						u[j++] = ef->upcase[i];
+					} else {
+						k = le16_to_cpu(ef->upcase[i+1]);
+						i++;
+						while (k--) {
+							le16_t v = cpu_to_le16(j);
+							u[j++] = v;
+						}
+					}
+				}
+				ef->upcase_chars = ucsize;
+				free(ef->upcase);
+				ef->upcase = u;
+			}
 			break;
```

**Comment by relan** [comment]
Thanks for the patch! I didn't know about this feature. How can I make an FS image with compressed upcase table? I need it to run tests.

**Comment by eiselekd** [comment]
I get it on an embedded system (it is not on the market), it is part of the exfat basic specification, chapter 7.2.5. Up-case Table.

**Comment by relan** [comment]
I've found that standard `format` utility from Windows makes FS with compressed upcase table. I'll fix it soon. Thanks again for reporting this issue!

**Comment by eiselekd** [comment]
great

**Comment by Agile86** [comment]
@relan can you tell how to produce exFAT with compressed table using standart Windows format tool?
(Is there some secret option or something).

**Comment by relan** [comment]
> can you tell how to produce exFAT with compressed table using standard Windows format tool?
(Is there some secret option or something).

No secret options required, `format` creates compressed upcase table by default. I've added support in 22e24732 (see the `upcomp` branch) but have not made a released yet.

**Comment by relan** [comment]
Fixed in v1.2.4. Sorry for the delay.
